model/riscv_vmem.sail

```diff
@@ -0,0 +1,849 @@
+                  let pa  = ((ppn << PAGESIZE_BITS) | offset_of_va(va));
+                  PTW_Success(pa, pte, pte_addr, level, global', ext_ptw)
+                }
+              }
+              else {
+                let pa = ((ppns << PAGESIZE_BITS) | offset_of_va (va));
+                PTW_Success(pa, pte, pte_addr, level, global', ext_ptw)
+              }
+          }
+        }
+      }
+    }
+  }
+}
+
+// ****************************************************************
+// Architectural SATP CSR
+
+// PUBLIC: see also riscv_insts_zicsr.sail and other CSR-related files
+register satp : xlenbits
+
+// See riscv_sys_regs.sail for legalize_satp{32,64}().
+// WARNING: those functions legalize Mode but not ASID?
+// PUBLIC: invoked from writeCSR() to fixup WARL fields
+function legalize_satp(a : Architecture,
+                       o : xlenbits,        // previous value of satp
+                       v : xlenbits)        // proposed new value of satp
+                      -> xlenbits = {       // new legal value of satp
+  if sizeof(xlen) == 32 then {
+    // The slice and extend ops below are no-ops when xlen==32,
+    // but appease the type-checker when xlen==64 (when this code is not executed!)
+    let o32      : bits(32) = o[31 .. 0];
+    let v32      : bits(32) = v[31 .. 0];
+    let new_satp : bits(32) = legalize_satp32(a, o32, v32);
+    zero_extend(new_satp);
+  } else if sizeof(xlen) == 64 then {
+    // The extend and truncate ops below are no-ops when xlen==64,
+    // but appease the type-checker when xlen==32 (when this code is not executed!)
+    let o64      : bits(64) = zero_extend(o);
+    let v64      : bits(64) = zero_extend(v);
+    let new_satp : bits(64) = legalize_satp64(a, o64, v64);
+    truncate(new_satp, sizeof(xlen))
+  } else
+    internal_error(__FILE__, __LINE__, "Unsupported xlen" ^ string_of_int(sizeof(xlen)))
+}
+
+// ----------------
+// Fields of SATP
+
+// ASID is 9b in Sv32, 16b in Sv39/Sv48/Sv57: we use 16b for both
+// PRIVATE
+function satp_to_asid(sv_params : SV_Params, satp_val : xlenbits) -> bits(16) = {
+  // This extend op is a no-op when xlen==64, extends when xlen==32
+  let satp_64b : bits(64) = zero_extend(satp_val);
+  let mask_64b : bits(64) = zero_extend(ones(sv_params.satp_asid_size_bits));
+  let asid_64b : bits(64) = (satp_64b >>  sv_params.satp_asid_lsb_index) & mask_64b;
+  asid_64b[15 .. 0]
+}
+
+// Result is 64b to cover both RV32 and RV64 addrs
+// PRIVATE
+function satp_to_PT_base(sv_params : SV_Params, satp_val : xlenbits) -> bits(64) = {
+  // This extend op is a no-op when xlen==64, extends when xlen==32
+  let satp_64b : bits(64) = zero_extend(satp_val);
+  let mask_64b : bits(64) = zero_extend(ones(sv_params.satp_ppn_size_bits));
+  let ppn_64b  : bits(64) = (satp_64b >>  sv_params.satp_ppn_lsb_index) & mask_64b;
+  ppn_64b << PAGESIZE_BITS
+}
+
+// Compute address translation mode from SATP
+// TODO: shouldn't we look at mstatus_UXL if priv is User?
```

Comment:
Since the format of `satp` is dependend on SXLEN and not UXLEN, why would this be?
Furthermore, the spec. explicitly states that "When SXLEN=32, the only valid setting for MODE is Sv32" and " When SXLEN=64, three paged virtual-memory schemes are defined: Sv39, Sv48, Sv57".

Reply:
Ok, thanks for the clarification.  I’ll remove the TODO in the comment.